File: backend/src/backend/query/ingest_failures.py
"""The failed-ingests read (issue 357): one scanner's rejected pushes, newest first, paged.

Paged by `search_after` with NO point-in-time. The series is append-only and read newest-first on
a total order (`@timestamp`, then the unique `failure_id`), so a record written mid-walk lands
AHEAD of the cursor and never shifts a later page — the stability a PIT would buy, without
spending the per-principal PIT budget on a panel that sits on a screen next to others. The one
thing that can move under a walk is retention dropping an old index, which only ends it early.

The window is the shared trend vocabulary (`days` + optional `as_of`, day-floored like every
trend), filtered on the server stamp. Per-scanner is sacred: `scanner` is required, so every page
and every total belongs to exactly one scanner.
"""

import base64
import binascii
import json
from datetime import datetime
from typing import Any

from backend.query.trends import window_bounds

SORT: list[dict[str, str]] = [{"@timestamp": "desc"}, {"failure_id": "desc"}]
ROW_FIELDS = [
    "@timestamp",
    "failure_id",
    "scanner",
    "stage",
    "reason",
    "status",
    "error",
    "image_ref",
]


def encode_cursor(sort_values: list[Any]) -> str:
    return base64.urlsafe_b64encode(json.dumps({"a": sort_values}).encode()).decode()


def decode_cursor(cursor: str) -> list[Any]:
    """The last row's sort values, shape-checked so a tampered cursor is a 422, never a 500
    from inside the search."""
    try:
        after = json.loads(base64.urlsafe_b64decode(cursor.encode()))["a"]
    except (binascii.Error, json.JSONDecodeError, KeyError, TypeError, ValueError) as exc:
        raise ValueError("invalid cursor") from exc
    if not (
        isinstance(after, list)
        and len(after) == 2
        and isinstance(after[0], int)
        and not isinstance(after[0], bool)
        and isinstance(after[1], str)
    ):
        raise ValueError("invalid cursor")
    return after


def build_ingest_failures_body(
    *,
    scanner: str,
    days: int,
    size: int,
    anchor: datetime | None = None,
    search_after: list[Any] | None = None,
) -> dict[str, Any]:
    """Pure builder. Asks for `size + 1` rows: the extra one only says whether a next page
    exists, so the last page never costs an empty round-trip."""
    gte, _ = window_bounds(days, anchor)
    window: dict[str, Any] = {"gte": gte}
    if anchor is not None:  # the ≤ T cut: a rewound read must not see later rejections
        window["lte"] = anchor.isoformat()
    body: dict[str, Any] = {
        "size": size + 1,
        "track_total_hits": True,
        "query": {
            "bool": {
                "filter": [
                    {"term": {"scanner": scanner}},
                    {"range": {"@timestamp": window}},
                ]
            }
        },
        "sort": SORT,
        "_source": ROW_FIELDS,
    }
    if search_after is not None:
        body["search_after"] = search_after
    return body


def page_of(resp: dict[str, Any], size: int) -> dict[str, Any]:
    """The list envelope (api-design.md, wrapped total): `data`, `total`, `next_cursor`."""
    hits = resp["hits"]["hits"]
    rows = hits[:size]
    more = len(hits) > size
    return {
        "data": [h["_source"] for h in rows],
        "total": resp["hits"]["total"],
        "next_cursor": encode_cursor(rows[-1]["sort"]) if more and rows else None,
    }

File: frontend/src/system/ingestFailures.ts
/**
 * The failed-ingests read (`GET /api/v1/scanners/ingest-failures`): its query params and row
 * shape. The window is the trend vocabulary (`days` + `as_of` only at T<now), so it is built by
 * the same `buildTrendQuery` every trend lens uses — one rounding rule for "the range".
 */
import { buildTrendQuery } from '@/charts/buildTrendQuery'

export type ScannerName = 'trivy' | 'grype'

export interface IngestFailureRow {
  '@timestamp': string
  failure_id: string
  scanner: string
  stage: string
  reason: string
  status: number
  error: string
  image_ref: string | null
}

export interface IngestFailuresPage {
  data: IngestFailureRow[]
  total: { value: number; relation: string }
  next_cursor: string | null
}

export function buildIngestFailuresQuery(p: {
  clusterId: string
  scanner: ScannerName
  windowDays: number
  t: string | null
  size: number
  cursor: string | null
}) {
  return {
    ...buildTrendQuery(p.clusterId, p.windowDays, p.t),
    scanner: p.scanner,
    size: p.size,
    ...(p.cursor === null ? {} : { cursor: p.cursor }),
  }
}
